### ISSUE TITLE
BertEncoder with and without flash attention doesn't match

### ISSUE THREAD

**Issue by jamaliki** [Dao-AILab/flash-attention#473]
Hi,

There seems to be a rather large difference between the BertEncoder with and without flash attention enabled. I might be making a mistake, please let me know.

Here is some minimal code:

```
import torch
from transformers import BertConfig
from flash_attn.models.bert import BertEncoder


config = BertConfig.from_pretrained("bert-base-uncased")

config.hidden_act = "gelu_new"
config.use_flash_attn = False
config.fused_bias_fc = True
config.fused_mlp = True
config.fused_dropout_add_ln = True
config.dense_seq_output = True
config.last_layer_subset = False
config.use_xentropy = True

model_weights = BertEncoder(config).state_dict()

dtype = torch.float16
batch_size = 4
max_seqlen = 512
seqlens = torch.randint(max_seqlen // 2, max_seqlen + 1, (batch_size,), device="cuda")
x = torch.rand(
    batch_size, max_seqlen, config.hidden_size, device="cuda", dtype=dtype
)


def test_bert_dense_seq_output(use_flash_attention):
    torch.manual_seed(0)
    dtype = torch.float16
    config = BertConfig.from_pretrained("bert-base-uncased")
    config.hidden_act = "gelu_new"
    config.use_flash_attn = use_flash_attention
    config.fused_bias_fc = True
    config.fused_mlp = True
    config.fused_dropout_add_ln = True
    config.dense_seq_output = True
    config.last_layer_subset = False
    config.use_xentropy = True

    model = BertEncoder(config)
    model = model.cuda().to(dtype=dtype)
    model.load_state_dict(model_weights)

    out = model(x)
    return out

out_flash = test_bert_dense_seq_output(True)
out_no_flash = test_bert_dense_seq_output(False)

print(torch.norm(out_flash - out_no_flash, dim=-1))
print(torch.isclose(out_flash, out_no_flash, rtol=0.0001, atol=1e-6))
```

The results for the norm are all large (around 11).

**Comment by jamaliki** [comment]
I have also just reproduced this with all the other optimizations turned off as well. Still does not match.

**Comment by tridao** [comment]
You want to compare
(reference model in fp32 - reference model in fp16) vs (reference model in fp32 - model with flash in fp16).
What are the two difference there?
We usually see the model with flash being more accurate (closer to the reference model in fp32).

**Comment by tridao** [comment]
If the differences are very large, then it's possible some options are mis-configured. You can take a look at our test for BERT [here](https://github.com/Dao-AILab/flash-attention/blob/main/tests/models/test_bert.py).

**Comment by tridao** [comment]
The model has dropout. You need to call `model.eval()` otherwise the results are always random each time.
Once you set `model.eval()` you'll see that the outputs are much closer.

**Comment by jamaliki** [comment]
@tridao thank you, that took care of the problem. Although I thought setting the seed would have made sure dropout was giving the same result? Either way, we can close this.

**Comment by rahul-k01** [comment]
@jamaliki, is your problem solved? I mean, have you implemented the flash attention successfully?
